modules/src/events.rs

```diff
@@ -0,0 +1,248 @@
+use crate::ics02_client::events as ClientEvents;
+use crate::ics02_client::events::NewBlock;
+use crate::ics03_connection::events as ConnectionEvents;
+use crate::ics04_channel::events as ChannelEvents;
+use crate::ics20_fungible_token_transfer::events as TransferEvents;
+use serde_derive::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::convert::{TryFrom, TryInto};
+use tendermint::block;
+use tendermint::rpc::event_listener::{ResultEvent, TMEventData};
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub enum IBCEvent {
+    NewBlock(NewBlock),
+
+    CreateClient(ClientEvents::CreateClient),
+    UpdateClient(ClientEvents::UpdateClient),
+    ClientMisbehavior(ClientEvents::ClientMisbehavior),
+
+    OpenInitConnection(ConnectionEvents::OpenInit),
+    OpenTryConnection(ConnectionEvents::OpenTry),
+    OpenAckConnection(ConnectionEvents::OpenAck),
+    OpenConfirmConnection(ConnectionEvents::OpenConfirm),
+
+    OpenInitChannel(ChannelEvents::OpenInit),
+    OpenTryChannel(ChannelEvents::OpenTry),
+    OpenAckChannel(ChannelEvents::OpenAck),
+    OpenConfirmChannel(ChannelEvents::OpenConfirm),
+    CloseInitChannel(ChannelEvents::CloseInit),
+    CloseConfirmChannel(ChannelEvents::CloseConfirm),
+
+    SendPacketChannel(ChannelEvents::SendPacket),
+    ReceivePacketChannel(ChannelEvents::ReceivePacket),
+    AcknowledgePacketChannel(ChannelEvents::AcknowledgePacket),
+    CleanupPacketChannel(ChannelEvents::CleanupPacket),
+    TimeoutPacketChannel(ChannelEvents::TimeoutPacket),
+
+    TimeoutTransfer(TransferEvents::Timeout),
+    PacketTransfer(TransferEvents::Packet),
+    ChannelClosedTransfer(TransferEvents::ChannelClosed),
+}
+
+impl IBCEvent {
+    pub fn to_json(&self) -> String {
+        serde_json::to_string(self).unwrap()
+    }
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct TryObject {
+    pub height: block::Height,
+    pub action: String,
+    pub idx: usize,
+    pub events: HashMap<String, Vec<String>>,
+}
+
+impl TryObject {
+    pub fn new(
+        height: block::Height,
+        action: String,
+        idx: usize,
+        events: HashMap<String, Vec<String>>,
+    ) -> TryObject {
+        TryObject {
+            height,
+            action,
+            idx,
+            events,
+        }
+    }
+}
+
+pub fn extract_events<'a, S: ::std::hash::BuildHasher>(
```

Comment:
Is this explicit lifetime actually needed here? I would expect rustc to be able to infer it.

Reply:
I added it because the compiler was complaining without it...but i can remove the lifetime once I change the return value to (). Here is the old error:
```
   Compiling relayer-modules v0.1.0 (/Users/ancaz/rust/2ibc-rs/modules)
error[E0106]: missing lifetime specifier
  --> modules/src/events.rs:76:13
   |
74 |     events: &HashMap<String, Vec<String>, S>,
   |             --------------------------------
75 |     action_string: &str,
   |                    ----
76 | ) -> Result<&HashMap<String, Vec<String>, S>, Box<dyn std::error::Error>> {
   |             ^ expected named lifetime parameter
   |
   = help: this function's return type contains a borrowed value, but the signature does not say whether it is borrowed from `events` or `action_string`
help: consider introducing a named lifetime parameter
```